Intrusive SHAMap smart pointers for efficient memory use and lock-free synchronization

vlntb opened the pull request:
## High Level Overview of Change

This PR finalises the work authored by Scott Determan (https://github.com/seelabs) and is based on the original PR (https://github.com/XRPLF/rippled/pull/4815).


### Context of Change

There are two goals:
- Synchronise this change with the most recent `develop` branch.
- Address outstanding questions raised in the original PR.

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [x] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.

For performance-impacting changes, please provide these details:
1. Is this a new feature, bug fix, or improvement to existing functionality?
2. What behavior/functionality does the change impact?
3. In what processing can the impact be measured? Be as specific as possible - e.g. RPC client call, payment transaction that involves LOB, AMM, caching, DB operations, etc.
4. Does this change affect concurrent processing - e.g. does it involve acquiring locks, multi-threaded processing, or async processing?
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

HowardHinnant reviewed (approved):
The theoretical maximal value for string references is calculated to be 210.  Experimental evidence also from the readme detects a value above the theoretical maximum: 387.
I ran a server for about an hour and detected a max of 1908.

These are all well below the limits of 65535, so this limit is probably safe.  But it wouldn't hurt to revisit the theoretical maximum and discover why it is incorrect.

vlntb commented:
> The theoretical maximal value for string references is calculated to be 210. Experimental evidence also from the readme detects a value above the theoretical maximum: 387. I ran a server for about an hour and detected a max of 1908.
> 
> These are all well below the limits of 65535, so this limit is probably safe. But it wouldn't hurt to revisit the theoretical maximum and discover why it is incorrect.


I did additional digging following a comment from @HowardHinnant. What I didn't take into account is that `rippled` is processing transaction or ledger data in a concurrent environment. I identified four types of routines that can happen in parallel:

- InboundLedgersImp::gotLedgerData
- SHAMap::walkTowardsKey
- SHAMap::flushDirty 
- LedgerMaster::gotFetchPack


Two of those routines are executed from the `JobQueue` and can be parallized further based on the node_size configuration parameter. The difference in this parameter explains the difference in the result that Howard and I received. Howard had his node_size set as `huge`, resulting in 8 threads in the JobQueue pool. In my setup, I had it defined as `medium`, resulting in 4 threads.

Based on those findings, we should update the Theoretical Maximum value. 
1) Theoretical Maximum value for a single thread: 210 = 2 X (15 / 5) X 35.
2) InboundLedgersImp::gotLedgerData can be executed across a maximum of 8 threads.
3) SHAMap::walkTowardsKey - 1 thread
4) SHAMap::flushDirty - 1 thread
5) LedgerMaster::gotFetchPack - can be executed across a maximum of 8 threads.

Giving overall Theoretical Maximum value as: 
8 x 210 +  210 + 210 + 8 x 210 = **3780**.

This is still significantly lower than the allocated **65535** range.

vlntb commented:
See the original PR and review comments here: 
https://github.com/XRPLF/rippled/pull/4815

vvysokikh1 commented on src/xrpld/shamap/TreeNodeCache.h:
```diff
@@ -21,10 +21,17 @@
+    SHAMapTreeNode,
+    /*IsKeyCache*/ false,
+    SharedWeakUnion<SHAMapTreeNode>,
+    SharedIntrusive<SHAMapTreeNode>>;
```
nit: should be intr_ptr::SharedPtr here as well?

Also I see there's no SharedWeakUnion exposed the same way. I wonder if that should?

vvysokikh1 commented on src/xrpld/shamap/detail/SHAMapInnerNode.cpp:
```diff
@@ -44,6 +45,17 @@ SHAMapInnerNode::SHAMapInnerNode(
+SHAMapInnerNode::partialDestructor()
+{
+    intr_ptr::SharedPtr<SHAMapTreeNode>* children;
+    // structured bindings can't be captured in c++ 17; use tie instead
```
nit: I suppose we can use it now?

vlntb replied:
Updated here + 3 other places to use modern structured bindings.

vvysokikh1 reviewed (commented):
not a complete review, mostly nits so far.

vlntb commented on include/xrpl/basics/IntrusivePointer.h:
```diff
@@ -0,0 +1,515 @@
+    operator=(SharedIntrusive<TT> const& rhs);
+
+    /** Adopt the raw pointer and increment the weak count. */
+    void
+    adopt(T* ptr);
```
Theoretically, we can have a case where we are transitioning from standard smart pointers to intr_ptr in stages, as we plan to continue integrating intr_ptr into other parts of rippled. During this transition, it is possible that some parts of the logic will already be using intr_ptr, while other parts will still rely on std::shared_ptr.

vvysokikh1 replied:
This makes sense for shared ptrs, but not for weak pointer. I'm just struggling to understand the use case of adopt for weak ptr (std weak_ptr has not such functionality either)

vlntb commented on include/xrpl/basics/IntrusivePointer.ipp:
```diff
@@ -0,0 +1,717 @@
+            ptr_->partialDestructor();
+            partialDestructorFinished(&ptr_);
+            // ptr_ is null and may no longer be used
+            break;
```
Looking through the rest of the codebase where `UNREACHABLE` is used, it appears that we prefer to future-proof the implementation. In this case, the inclusion of `partialDestructor` provides a safeguard should a situation arise where partial destruction becomes legitimate for a weak pointer.

vvysokikh1 replied:
I would agree with this statement most of the time, but the whole intent and idea of intrusive pointer here is that weak pointer must not call partial destruction. So even if such need suddenly arises, this whole idea and implementation will have to be revamped.

vlntb replied:
Resolved by introducing separate ReleaseStrongRefAction and ReleaseWeakRefAction.

vlntb commented on include/xrpl/basics/IntrusivePointer.ipp:
```diff
@@ -0,0 +1,717 @@
+            assert(0);
+            delete p;
+            unsafeSetRawPtr(nullptr);
+            return true;  // Should never happen
```
Same as above, the pattern is to future-proof the implementation, but I don't have strong feelings about this approach and happy to remove the dead code.

vlntb replied:
Resolved by introducing separate ReleaseStrongRefAction and ReleaseWeakRefAction.

vlntb commented on include/xrpl/basics/IntrusiveRefCounts.h:
```diff
@@ -0,0 +1,466 @@
+{
+    using enum ReleaseRefAction;
+
+    static_assert(weakDelta > strongDelta);
```
I understand that Scott's idea was to perform the check where it truly matters. In the future, the nature of those constants might change, and they could become variables.

vlntb commented on src/xrpld/shamap/SHAMapInnerNode.h:
```diff
@@ -148,19 +153,21 @@ class SHAMapInnerNode final : public SHAMapTreeNode,
+        requires std::derived_from<T, SHAMapTreeNode>
     void
-    shareChild(int m, std::shared_ptr<SHAMapTreeNode> const& child);
+    shareChild(int m, SharedIntrusive<T> const& child);
```
I agree. Well spotted!

vvysokikh1 commented on include/xrpl/basics/IntrusivePointer.h:
```diff
@@ -0,0 +1,519 @@
+    operator=(SharedIntrusive&& rhs);
+
+    template <class TT>
+    // clang-format off
```
nit: i tried to remove clang-format off from here (and place above). It doesn't seem to change the way it's formatted. Do you know if the previous version of clang-format was messing this up? It doesn't seem like this is required now.

vlntb replied:
It does one minor thing: `requires` stays aligned with the rest of the template definition.


```
template <class T>
template <class TT>
    requires std::convertible_to<TT*, T*>
SharedIntrusive<T>&
SharedIntrusive<T>::operator=(SharedIntrusive<TT>&& rhs)
{
    static_assert(
        !std::is_same_v<T, TT>,
        "This overload should not be instantiated for T == TT");

    unsafeReleaseAndStore(rhs.unsafeExchange(nullptr));
    return *this;
}
```

vvysokikh1 replied:
ok, whatever approach we take that should be uniform.

lines 103 and 109 do not use clang format off, so please add it there or remove it here :)

vlntb replied:
I did a bit more digging into the question of leading space since we are setting an example here for consistent style. At least two reputable C++ codebases [Boost](https://github.com/boostorg/parser/blob/57cdd78210432c726d3e4f3969b7a39317c1b9f2/include/boost/parser/detail/text/transcode_iterator_fwd.hpp#L42) and  [LLVM](https://github.com/llvm/llvm-project/blob/579ced4f8266b273d15b2801067a828151a222ef/libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/requirements.compile.pass.cpp#L197) are using space at the beginning of new line for multiline definitions:
ie:
```
template<
        format FromFormat,
        format ToFormat,
        std::input_iterator I,
        std::sentinel_for<I> S = I,
        transcoding_error_handler ErrorHandler = use_replacement_character>
        requires std::convertible_to<std::iter_value_t<I>, detail::format_to_type_t<FromFormat>>
```
and

```
template <class T>
  requires(!std::convertible_to<T, int>)
void requires_init_is_convertible_to_decayed() {
  static_assert(!requires(std::ranges::subrange<int*, int*> r, T init) {
    std::ranges::fold_left_with_iter(r.begin(), r.end(), init, std::plus());
 });
```

This is what our clang-format config suggests as well. In my view, overriding this style with clang-format off/on only makes the code less readable. Removing clang format overrides ...

vvysokikh1 commented on include/xrpl/basics/IntrusivePointer.h:
```diff
@@ -0,0 +1,519 @@
+template <class T>
+class SharedWeakUnion
+{
+    static_assert(
+        alignof(T) >= 2,
+        "Bad alignment: Combo pointer requires low bit to be zero");
+
```
nit: I would suggest adding comment from line 430 here. That could improve readability

vvysokikh1 reviewed (approved):
Please add a bit more tests for coverage and it's lgtm

legleux commented:
Should these flags be defined in the top level `CMakeLists.txt` instead of `cmake/RippledCompiler.cmake`? Eventually, I'd like to move all this to `CMakePresets.json` anyway.

`-Wno-subobject-linkage` produces
```
cc1: warning: command-line option ‘-Wno-subobject-linkage’ is valid for C++/ObjC++ but not for C
```

nit: Why check each compiler if they're mutually exclusive (if/if/if)?

nit: the `if (` i.e. <CMakeFunction>[::space::]<parens> is all over the CMake for `rippled` but the docs and most newer CMake specify _no space_ so I've been trying to stick to that and slowly convert it everywhere I see it.

vlntb commented:
> Should these flags be defined in the top level `CMakeLists.txt` instead of `cmake/RippledCompiler.cmake`? Eventually, I'd like to move all this to `CMakePresets.json` anyway.
> 
> `-Wno-subobject-linkage` produces
> 
> ```
> cc1: warning: command-line option ‘-Wno-subobject-linkage’ is valid for C++/ObjC++ but not for C
> ```
> 
> nit: Why check each compiler if they're mutually exclusive (if/if/if)?
> 
> nit: the `if (` i.e. [::space::] is all over the CMake for `rippled` but the docs and most newer CMake specify _no space_ so I've been trying to stick to that and slowly convert it everywhere I see it.

* The reason for having those flags on the current level is that other `add_compile_options` are already defined there. We can, in the future, move everything to the level up, but we should do this for everything at once.
* Addressed exclusivity comment - moved to if-elseif-ifelse.
* Removed space after `if` and an empty line to follow the convention.

vlntb commented on include/xrpl/basics/TaggedCache.h:
```diff
@@ -70,230 +76,55 @@ class TaggedCache
-
-        JLOG(m_journal.debug()) << m_name << " target size set to " << s;
-    }
+    setTargetSize(int s);
```
Agree, we only initialise those variables through constructor and never update afterwards. Makes sense to make it const and remove setters.

vlntb commented on include/xrpl/basics/TaggedCache.h:
```diff
@@ -70,230 +76,55 @@ class TaggedCache
-        std::lock_guard lock(m_mutex);
-        return m_target_age;
-    }
+    getTargetAge() const;
```
Yes, same for getters

vlntb commented on src/xrpld/nodestore/Database.h:
```diff
@@ -26,6 +26,7 @@
 
 #include <xrpl/basics/BasicConfig.h>
 #include <xrpl/basics/Log.h>
+#include <xrpl/basics/TaggedCache.ipp>
```
Checked everything with iwyu_tool.py tool

vlntb commented on src/xrpld/shamap/SHAMapTreeNode.h:
```diff
@@ -23,6 +23,9 @@
 #include <xrpld/shamap/SHAMapItem.h>
 #include <xrpld/shamap/SHAMapNodeID.h>
 
+#include <xrpl/basics/CountedObject.h>
```
Checked everything with iwyu_tool.py tool

vlntb commented on src/xrpld/ledger/detail/CachedView.cpp:
```diff
@@ -19,6 +19,10 @@
+#include <xrpl/basics/TaggedCache.ipp>
+#include <xrpl/basics/contract.h>
+#include <xrpl/protocol/Serializer.h>
+
```
Checked everything with iwyu_tool.py tool

vlntb commented on include/xrpl/basics/TaggedCache.ipp:
I agree that there is no ideal solution here — each approach involves a different type of compromise.

The alternative to separation is to keep everything in a single `TaggedCache.h` file. While this would save some lines of code, it would reduce the clarity of the `TaggedCache` interface.

I would vote in favour of future readability, even at the cost of some current complications, and support reviewing the splitting change.
